### PR TITLE
feat: KMP matcher

### PR DESCRIPTION
**Pull request by fgdorais**
- [x] Depends on #219 

A versatile implementation of the Knuth-Morris-Pratt matching algorithm.

### CODE REVIEW

**Comment by fgdorais** [comment]
Now integrates #172

**Comment by fgdorais** [comment]
Good idea! Do you think `@[specialize]` might be enough? Or do you absolutely need a reusable matcher type as in the Array API?

**Comment by fgdorais** [comment]
I added a few things that you might find useful. This API would be a lot simpler if there was a coercion from `String` to `Substring`.

**Comment by nomeata** [comment]
Great, thanks!

> This API would be a lot simpler if there was a coercion from `String` to `Substring`.

Could there be one?

File: Std/Data/String/Basic.lean [hunk 1]
```diff
@@ -1,15 +1,63 @@
...
 -/
 
 import Std.Data.Char
 import Std.Data.Nat.Lemmas
+import Std.Data.Array.Match
 
-protected theorem String.Pos.ne_zero_of_lt : {a b : Pos} → a < b → b ≠ 0
+namespace String
+
+instance : Coe String Substring := ⟨toSubstring⟩
```
Comment (digama0) [inline Std/Data/String/Basic.lean:13]: This should be a separate PR, which also removes most uses of `toSubstring` in the rest of the code at the same time

File: Std/Data/String/Basic.lean [hunk 2]
```diff
@@ -1,15 +1,63 @@
...
+  loop s #[]
+where
+  /-- Accumulator loop for `String.Matcher.findAll` -/
+  loop (s : Substring) (occs : Array Substring) : Array Substring :=
+    match m.next? s with
+    | none => occs
+    | some (s, _) =>
+      loop s <| occs.push {s with
+        startPos := ⟨s.startPos.byteIdx - m.patternSize⟩
+        stopPos := s.startPos}
```
Comment (digama0) [inline Std/Data/String/Basic.lean:48]:
```suggestion
      loop s <| occs.push { s with
        startPos := ⟨s.startPos.byteIdx - m.patternSize⟩
        stopPos := s.startPos }
```

File: Std/Data/String/Basic.lean [hunk 3]
```diff
@@ -72,10 +120,41 @@ def dropSuffix? (s : Substring) (suff : Substring) : Option Substring :=
   else
     none
 
+/--
+Returns all the substrings of `s` that martch `pattern`.
```
Comment (digama0) [inline Std/Data/String/Basic.lean:124]:
```suggestion
Returns all the substrings of `s` that match `pattern`.
```

File: Std/Data/String/Basic.lean [hunk 4]
```diff
@@ -1,15 +1,63 @@
...
 
-protected theorem String.Pos.ne_zero_of_lt : {a b : Pos} → a < b → b ≠ 0
+namespace String
+
+instance : Coe String Substring := ⟨toSubstring⟩
+
+protected theorem Pos.ne_zero_of_lt : {a b : Pos} → a < b → b ≠ 0
 | _, _, hlt, rfl => Nat.not_lt_zero _ hlt
 
+/-- KMP Matcher -/
```
Comment (digama0) [inline Std/Data/String/Basic.lean:18]: KMP needs to be explained here (it should be okay on functions as long as the type explains it, but generally you shouldn't assume too much context in docstrings)

Comment (digama0) [reply Std/Data/String/Basic.lean:18]: Actually this is probably a good place to put the overall documentation on how to use this type effectively, maybe put a paragraph and a code block showing basic usage.

File: Std/Data/String/Basic.lean (hunk 1 above)
Comment (fgdorais) [reply Std/Data/String/Basic.lean:13]: #219

File: Std/Data/String/Basic.lean (hunk 4 above)
Comment (fgdorais) [reply Std/Data/String/Basic.lean]: I found an embarrassing bug so I added a basic test suite as well.

Comment (fgdorais) [reply Std/Data/String/Basic.lean]: What do you think about the description? Should I add something about how the basic working of the algorithm? (I don't think it's necessary for users to know how it works but this is a neat one.)

Comment (digama0) [reply Std/Data/String/Basic.lean]: The docs here can just reference the algorithm, possibly including a link to e.g. the wikipedia page. For a more detailed description of the algorithm you should use the module doc on `Data.Array.Match`.

File: Std/Data/Array/Match.lean [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+abbrev PrefixTable.pattern (t : PrefixTable α) : Array α := t.toArray.map Prod.fst
+
+/-- Transition function for the KMP matcher
+
+  Assuming we have an input `xs` with a suffix that matches the pattern prefix `t.pattern[:len]`
+  where `len : Fin (t.size+1)`. Then `xs.push x` has a suffix that matches the pattern prefix
+  `t.pattern[:t.step x len]`. If `len` is as large as possible then `t.step x len` will also be
+  as large as possible.
+-/
+def PrefixTable.step [BEq α] (t : PrefixTable α) (x : α) : Fin (t.size+1) → Fin (t.size+1)
```
Comment (semorrison) [inline Std/Data/Array/Match.lean:36]: Could `[BEq \a]` be a variable?

File: Std/Data/Array/Match.lean [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+    | some (x, stream) => loop (t.extend x) stream
+
+/-- KMP matcher structure -/
+structure Matcher (α) where
+  /-- Prefix table for the pattern -/
+  table : PrefixTable α
+  /-- Current longest matching prefix -/
+  state : Fin (table.size+1) := 0
+
+/-- Make a KMP matcher for a given a pattern array -/
```
Comment (semorrison) [inline Std/Data/Array/Match.lean:102]:
```suggestion
/-- Make a KMP matcher for a given  pattern array -/
```

File: Std/Data/Array/Match.lean [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+def Matcher.ofArray [BEq α] (pat : Array α) : Matcher α where
+  table := mkPrefixTable pat
+
+/-- Make a KMP matcher for a given a pattern stream -/
+def Matcher.ofStream [BEq α] [Stream σ α] (pat : σ) : Matcher α where
+  table := mkPrefixTableOfStream pat
+
+/-- Find next match from a given stream
+
+  Runs the stream until it reads a sequence that mathes the sought pattern, then returns the stream
```
Comment (semorrison) [inline Std/Data/Array/Match.lean:112]:
```suggestion
  Runs the stream until it reads a sequence that matches the sought pattern, then returns the stream
```

File: Std/Data/Array/Match.lean (hunk 5 above)
Comment (fgdorais) [reply Std/Data/Array/Match.lean:36]: I'm not sure what you mean. Just making it an implicit parameter?

File: Std/Data/String/Basic.lean [hunk 8]
```diff
@@ -74,10 +140,41 @@ def dropSuffix? (s : Substring) (suff : Substring) : Option Substring :=
...
+Returns true iff `pattern` occurs as a substring of `s`.
+-/
+@[inline] def containsSubstr (s pattern : Substring) : Bool :=
+  s.findSubstr? pattern |>.isSome
+
 end Substring
 
 namespace String
 
+@[inherit_doc Substring.findSubstr?]
```
Comment (semorrison) [inline Std/Data/String/Basic.lean:166]:
```suggestion
@[inherit_doc Substring.findAllSubstr]
```

File: Std/Data/String/Basic.lean [hunk 9]
```diff
@@ -74,10 +140,41 @@ def dropSuffix? (s : Substring) (suff : Substring) : Option Substring :=
...
+abbrev findAllSubstr (s pattern : Substring) : Array Substring :=
```
Comment (semorrison) [inline Std/Data/String/Basic.lean:167]: `s : String`

File: test/kmp_matcher.lean [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+#guard_msgs in
+#eval Option.isSome <| m.find? "AbbabbA" -- false
+
+/-- info: true -/
+#guard_msgs in
+#eval Option.isSome <| m.find? "aabbaa" -- true
+
+/-- info: 2 -/
+#guard_msgs in
+#eval Array.size <| m.findAll "abbabba" -- 2
```
Comment (semorrison) [inline test/kmp_matcher.lean:19]:
```suggestion
#guard (Array.size <| m.findAll "abbabba") = 2
```
etc for the other tests.

Comment (fgdorais) [reply test/kmp_matcher.lean:19]: Will do.

File: test/kmp_matcher.lean [hunk 11]
```diff
@@ -0,0 +1,23 @@
+import Std.Tactic.GuardMsgs
+import Std.Data.String.Basic
+
+/-! Tests for Knuth-Morris-Pratt matching algorithm -/
```
Comment (semorrison) [inline test/kmp_matcher.lean:4]: Wouldn't hurt for the tests to directly exercise `String.findSubstr?`, given that's the main entrypoint.

File: Std/Data/String/Basic.lean [hunk 12]
```diff
@@ -1,17 +1,83 @@
...
+  size := m.patternSize
+  /-- Accumulator loop for `String.Matcher.findAll` -/
+  loop (s : Substring) (m : Matcher) (occs : Array Substring) : Array Substring :=
+    match m.next? s with
+    | none => occs
+    | some (s, m) =>
+      loop s m <| occs.push { s with
+        startPos := ⟨s.startPos.byteIdx - size⟩
+        stopPos := s.startPos }
+
```
Comment (semorrison) [inline Std/Data/String/Basic.lean:69]: Not a particularly serious suggestion, but you could produce a lazy list of matches here. :-)

File: Std/Data/Array/Match.lean (hunk 5 above)
Comment (semorrison) [reply Std/Data/Array/Match.lean:36]: Sorry: write `variable [BEq \a]` at the top of the file, so it isn't in every type signature.

File: Std/Data/Array/Match.lean [hunk 13]
```diff
@@ -0,0 +1,126 @@
...
+def Matcher.ofStream [BEq α] [Stream σ α] (pat : σ) : Matcher α where
+  table := mkPrefixTableOfStream pat
+
+/-- Find next match from a given stream
+
+  Runs the stream until it reads a sequence that matches the sought pattern, then returns the stream
+  state at that point and an updated matcher state.
+-/
+partial def Matcher.next? [BEq α] [Stream σ α] (m : Matcher α) (stream : σ) :
+  Option (σ × Matcher α) :=
```
Comment (semorrison) [inline Std/Data/Array/Match.lean:116]:
```suggestion
    Option (σ × Matcher α) :=
```
I suspect the same indenting rules apply here as in Mathlib: when the type signature continue over multiple lines it is indented 4 spaces.

File: Std/Data/String/Basic.lean (hunk 12 above)
Comment (fgdorais) [reply Std/Data/String/Basic.lean:69]: Good question! This is a definite possibility using `MLList`. I'll think about it. If it turns out to be straightforward I will include it here but a separate PR might be a better venue.

File: Std/Data/Array/Match.lean (hunk 5 above)
Comment (digama0) [reply Std/Data/Array/Match.lean]: @semorrison I try to avoid `variable` wherever possible in std. In this case I think it's fine as is.

Comment (fgdorais) [reply Std/Data/Array/Match.lean:36]: I just reverted the change but I must admit having no personal preference either way.

File: Std/Data/String/Basic.lean (hunk 12 above)
Comment (fgdorais) [reply Std/Data/String/Basic.lean:69]: I thought about it for a bit. I like the idea but I think this would belong in a later PR. There seems to be a few choices available (`MLList` and `Stream` are the top two). Also, `Array.Match.next?` is already there as a low-level infrastructure to generate matches lazily. Since you said this was not particularly serious, I will politely pass.

File: test/kmp_matcher.lean [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+
+/-- Matcher for pattern "abba" -/
+def m := String.Matcher.ofString "abba"
+
+#guard Option.isSome (m.find? "AbbabbA") = false
```
Comment (semorrison) [inline test/kmp_matcher.lean:9]:
```suggestion
#guard ! (m.find? "AbbabbA").isSome
```
etc?

**Review by semorrison — APPROVED** [review]
Otherwise, LGTM now.

File: Std/Data/String/Basic.lean [hunk 15]
```diff
@@ -1,17 +1,83 @@
...
+@[inline] def Matcher.ofString (pattern : String) : Matcher :=
+  Matcher.ofSubstring pattern
+
+/-- The string pattern for the matcher -/
+def Matcher.pattern (m : Matcher) : String :=
+  m.table.foldl (fun s (c,_) => s.push c) ""
+
+/-- The byte size of the string pattern for the matcher -/
+def Matcher.patternSize (m : Matcher) : Nat :=
+  String.Pos.byteIdx <| m.table.foldl (fun n (c,_) => n + c) 0
```
Comment (digama0) [inline Std/Data/String/Basic.lean:53]: It is surprising that this is not O(1), and I believe this leads to an unexpected overhead in `findAll` below, which calls it in every iteration. (You shouldn't use `let rec` for precomputed state! It's just another recursive function.)

Comment (fgdorais) [reply Std/Data/String/Basic.lean]: Yes, that's much better!

File: Std/Data/Array/Match.lean [hunk 16]
```diff
@@ -0,0 +1,126 @@
...
+  valid : (h : i < toArray.size) → toArray[i].2 ≤ i
+
+instance : Inhabited (PrefixTable α) where
+  default := ⟨#[], fun .⟩
+
+/-- Returns the size of the prefix table -/
+abbrev PrefixTable.size (t : PrefixTable α) := t.toArray.size
+
+/-- Returns the pattern array of the prefix table -/
+abbrev PrefixTable.pattern (t : PrefixTable α) : Array α := t.toArray.map Prod.fst
```
Comment (digama0) [inline Std/Data/Array/Match.lean:27]: This operation is not O(1), which I think messes up the asymptotic order of the algorithm because it is called in `PrefixTable.step`. Maybe the numbers array should just be a parallel array instead of being interleaved with the pattern?

File: Std/Data/Array/Match.lean [hunk 17]
```diff
@@ -35,24 +35,19 @@ abbrev PrefixTable.pattern (t : PrefixTable α) : Array α := t.toArray.map Prod
...
     | k + 1 =>
       have h2 : k < t.size := Nat.lt_of_succ_lt_succ hk
       let k' := t.toArray[k].2
       have hk' : k' < k + 1 := Nat.lt_succ_of_le (t.valid h2)
       step t x ⟨k', Nat.lt_trans hk' hk⟩
+  if hsz : k < t.size then
+    if x == t.toArray[k].1 then
+      ⟨k+1, Nat.succ_lt_succ hsz⟩
+    else cont ()
+  else cont ()
```
Comment (fgdorais) [inline Std/Data/Array/Match.lean:47]: This is cool! I didn't realize Lean would automatically inline `cont` so that this is still tail recursive.
